lib/improver/tests/utilities/test_cube_units.py

```diff
@@ -34,38 +34,296 @@
+        """Test error is raised when enforce=False"""
+        self.percentile_cube.coord('percentile').points = (
+            self.percentile_cube.coord('percentile').points.astype(np.int32))
+        msg = "does not conform"
+        with self.assertRaisesRegex(ValueError, msg):
+            cube_units.enforce_units_and_dtypes(
+                self.percentile_cube, enforce=False)
+
+    def test_coordinates_correctly_identified(self):
+        """Test all coordinates in a heterogeneous cube list are identified and
+        corrected"""
+        self.percentile_cube.coord('percentile').points = (
+            self.percentile_cube.coord('percentile').points.astype(np.int32))
+        self.probability_cube.coord(
+            'air_temperature').convert_units('Fahrenheit')
+        result = cube_units.enforce_units_and_dtypes(
+            [self.percentile_cube, self.probability_cube])
+        self.assertEqual(result[0].coord('percentile').dtype, np.float32)
+        self.assertEqual(result[1].coord('air_temperature').units, 'K')
+
+    def test_subset_of_coordinates(self):
+        """Test function can enforce on a selected subset of coordinates and
+        leave all others unchanged"""
+        self.percentile_cube.coord('percentile').points = (
+            self.percentile_cube.coord('percentile').points.astype(np.int32))
+        self.percentile_cube.coord('time').convert_units(
+            'hours since 1970-01-01 00:00:00')
+        self.probability_cube.coord(
+            'air_temperature').convert_units('Fahrenheit')
+        self.probability_cube.coord('forecast_period').convert_units('h')
+
+        result = cube_units.enforce_units_and_dtypes(
+            [self.percentile_cube, self.probability_cube],
+            coords=["time", "forecast_period"])
+        self.assertEqual(result[0].coord('percentile').dtype, np.int32)
+        self.assertEqual(
+            result[0].coord('time').units, 'seconds since 1970-01-01 00:00:00')
+        self.assertEqual(
+            result[1].coord('air_temperature').units, 'Fahrenheit')
+        self.assertEqual(result[1].coord('forecast_period').units, 's')
+
+    def test_quantity_unavailable(self):
+        """Test error raised if the named quantity is not listed in the
+        dictionary standard"""
+        self.data_cube.rename("number_of_fish")
+        self.data_cube.units = "1"
+        msg = "Name 'number_of_fish' is not uniquely defined in units.py"
+        with self.assertRaisesRegex(KeyError, msg):
+            cube_units.enforce_units_and_dtypes(self.data_cube)
+
+    def test_multiple_errors(self):
+        """Test a list of errors is correctly caught and re-raised"""
+        self.data_cube.convert_units('Fahrenheit')
+        self.probability_cube.coord(
+            'air_temperature').convert_units('Fahrenheit')
+        msg = ("The following errors were raised during processing:\n"
+               "air_temperature with units Fahrenheit and datatype ")
```

Comment:
If you want the complete message you need a few escape characters (\) on the brackets:

```
msg = ("The following errors were raised during processing:\n"      
       "air_temperature with units Fahrenheit and datatype float32 "
       "does not conform to expected standard \(units K, datatype " 
       "\<class 'numpy.float32'\>\)")                               
```

Reply:
Excellent - thank you.